src/cargo/sources/git/utils.rs

```diff
@@ -977,6 +977,13 @@ pub fn fetch(
                 // The reason we write to `refs/remotes/origin/HEAD` is that it's of special significance
                 // when during `GitReference::resolve()`, but otherwise it shouldn't matter.
                 refspecs.push(format!("+{0}:refs/remotes/origin/HEAD", rev));
+            } else if Url::parse(remote_url).map_or(false, |url| is_github(&url))
```

Comment:
Thank you for the nice finding and write-up!

I understand supporting other Git hosting would be awesome, though this is still a GitHub specific fast path. I wonder if we could move this patch to under [`github_fast_path`](https://github.com/rust-lang/cargo/blob/84dc5dc11a9007a08f27170454da6097265e510e/src/cargo/sources/git/utils.rs#L1420). Maybe something like this?

```diff
                if let Some(local_object) = local_object {
                    if is_short_hash_of(rev, local_object) {
                        debug!("github fast path already has {local_object}");
                        return Ok(FastPathRev::UpToDate);
                    }
                }
+               if let Ok(oid) = rev.parse::<Oid>() {
+                    debug!("github fast path is already a full commit hash {rev}");
+                    return Ok(FastPathRev::NeedsFetch(oid))
+               }
                rev
            } else {
                debug!("can't use github fast path with `rev = \"{}\"`", rev);
                return Ok(FastPathRev::Indeterminate);
            }
```

Reply:
Oh that's clever! Let me give that a shot.